ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs

```diff
@@ -364,6 +380,99 @@ broadcastNetwork registry btime numCoreNodes pInfo initRNG numSlots slotLen = do
         , cdbGcDelay          = 0
         }
 
+{-------------------------------------------------------------------------------
+  Core Node Topologies
+-------------------------------------------------------------------------------}
+
+-- | The topology of a network of core nodes
+--
+-- INVARIANT: The topology is a /connected simple graph/: no loops, no cycles,
+-- every node is reachable from every other.
+--
+-- This data type maps each node to the *lesser* nodes to which it is
+-- connected. This list is therefore inherently topologically sorted (a.k.a.
+-- sorted according to dependencies). By the above invariant, Node 0 must be
```

Comment:
Hmm. I was thinking ahead to #231 here... the invariant implies those requirements only if we assume that the topology is fully connected even if all nodes haven't joined yet and that nodes join in ascending order of Id.

Reply:
For example, 0 <-> 2 <-> 1 satisfies the invariant, but does not have a 0 <-> 1 edge. However, because our nodes all have the same behavior (ignoring `LeaderSchedule` etc), that graph is isomorphic to the graph 0 <-> 1 <-> 2 which does satisfy the invariant.